types/react/index.d.ts

```diff
@@ -2111,6 +1882,52 @@ declare namespace React {
         getServerSnapshot?: () => Snapshot,
     ): Snapshot;
 
+    export function useOptimistic<State>(
+        passthrough: State,
+    ): [State, (action: State | ((pendingState: State) => State)) => void];
+    export function useOptimistic<State, Action>(
+        passthrough: State,
+        reducer: (state: State, action: Action) => State,
+    ): [State, (action: Action) => void];
+
+    interface ThenableImpl<T> {
+        then(onFulfill: (value: T) => unknown, onReject: (error: unknown) => unknown): void | PromiseLike<unknown>;
+    }
+    interface UntrackedThenable<T> extends ThenableImpl<T> {
+        status?: void;
+    }
+
+    export interface PendingThenable<T> extends ThenableImpl<T> {
+        status: "pending";
+    }
+
+    export interface FulfilledThenable<T> extends ThenableImpl<T> {
+        status: "fulfilled";
+        value: T;
+    }
+
+    export interface RejectedThenable<T> extends ThenableImpl<T> {
+        status: "rejected";
+        reason: unknown;
+    }
+
+    export type Thenable<T> = UntrackedThenable<T> | PendingThenable<T> | FulfilledThenable<T> | RejectedThenable<T>;
+
+    export type Usable<T> = Thenable<T> | Context<T>;
```

Comment:
This looks overly complex. Also native promises don’t have the `status` property. Does React have some special kind of internal promise type? And are users supposed to deal with that? Why not use this?

```suggestion
    export type Usable<T> = PromiseLike<T> | Context<T>;
```

Reply:
Happy to look at this separately since it's already in the canary types.

Reply:
That makes sense. Do you want me to make a PR now or is this something for after React 19? Removing those exports could be considered breaking.

Reply:
Better now than never